### DISCUSSION TITLE
[HOW] stax log intermediate layer outputs

### DISCUSSION THREAD

**Discussion by llCurious** [jax-ml/jax#9221] [Q&A]
I have tried flax before and there is a way to capture the intermediate outputs following the link https://flax.readthedocs.io/en/latest/howtos/extracting_intermediates.html#use-capture-intermediates.

Recently, i try to use stax which i think is more friendly to customized layers. And i wonder if there is a way that i could log the intermediate outputs in each layer (or some layer) in a similar way?

**Comment by tomhennigan** [comment]
Stax is not actively developed any more and does not have this functionality (plumbing intermediate values) built in.

There is a separate library available called [Oryx](https://www.tensorflow.org/probability/oryx/notebooks/a_tour_of_oryx) which has a transformation called [Harvest](https://www.tensorflow.org/probability/oryx/api_docs/python/oryx/core/interpreters/harvest) which can be used with any JAX code (e.g. you could use this with Stax, Flax, whatever). This transformation allows you to capture intermediate values and and collect them at the end of your function.

**Reply by llCurious** [reply]
Got it. So i should tag values first and then `reap` them. However, if i write the following stax code, how could i obtain the output from each layer?

``` Python
nn_init, nn_apply = stax.serial(
        stax.Conv(2, (1, 1)),
        Flatten,
        stax.Dense(10),
        elementwise(stax.relu)
    )
```

**Reply by tomhennigan** [reply]
You would need to insert a stax layer that had no parameters and called harvest on the input. This is actually what `elementwise` does so you can probably use that (I've not tested this):

```python
nn_init, nn_apply = stax.serial(
        stax.Conv(2, (1, 1)),
        stax.elementwise(oryx.core.sow, name='layer_0', tag='intermediate'),
       ...
)
```

**Reply by llCurious** [reply]
Yep, the elementwise works fine.
However, it seems that the grad and reap cannot be done in the same time?
```Python
def loss_func(params, x, label):
    y = nn_apply(params, x)
    return mse_loss(y, label)

loss, gradients = jax.value_and_grad(loss_func)(opt_get_weights(opt_state), X, Y)
print('Reap:', oryx.core.reap(jax.value_and_grad(loss_func), tag='intermediate')(opt_get_weights(opt_state), X, Y))  # Pulls out 'layer_0'
```

**Reply by tomhennigan** [reply]
Seems like for value and reaped values you can use `oryx.core.harvest(..)` tensorflow/probability#1248 and some more discussion here tensorflow/probability#1045.

Perhaps @sharadmv has some thoughts?

**Reply by krzysztofrusek** [reply]
@tomhennigan is `sow/harvest` compatible with haiku transformations? 
Ideally, I would like to sow intermediates inside `hk.transform` and harvest them otside.
~Now I am using state for this, and oryx would be more elegant but I think it needs lifted sow aka~ ~`hk.vmap`~ ~etc.~ 

PS

It works !
her is [gist](https://gist.github.com/krzysztofrusek/9f46043f50ca4ef1ba7c7f3f545add60), though I am not sure if this is side effect free.

**Reply by llCurious** [reply]
thank you for your suggestion. The method you present works indeed, while the following approach is enough to sow and reap the intermediate values.

> You would need to insert a stax layer that had no parameters and called harvest on the input. This is actually what `elementwise` does so you can probably use that (I've not tested this):
> 
> ```python
> nn_init, nn_apply = stax.serial(
>         stax.Conv(2, (1, 1)),
>         stax.elementwise(oryx.core.sow, name='layer_0', tag='intermediate'),
>        ...
> )
> ```

What i am trying to do next is to obtain the intermediate, along with the the value and gradients which are more important. The current approach seems cannot obtain both simutaneously.
